evalml/tests/pipeline_tests/test_component_graph.py

```diff
@@ -0,0 +1,576 @@
+
+    def fit(self, X, y):
+        return self
+
+
+class EstimatorA(DummyEstimator):
+    """copy class"""
+
+
+class EstimatorB(DummyEstimator):
+    """copy class"""
+
+
+class EstimatorC(DummyEstimator):
+    """copy class"""
+
+
+@pytest.fixture
+def dummy_components():
+    return TransformerA, TransformerB, TransformerC, EstimatorA, EstimatorB, EstimatorC
+
+
+@pytest.fixture
+def example_graph():
+    graph = {'Imputer': [Imputer],
+             'OneHot_RandomForest': [OneHotEncoder, 'Imputer.x'],
+             'OneHot_ElasticNet': [OneHotEncoder, 'Imputer.x'],
+             'Random Forest': [RandomForestClassifier, 'OneHot_RandomForest.x'],
+             'Elastic Net': [ElasticNetClassifier, 'OneHot_ElasticNet.x'],
+             'Logistic Regression': [LogisticRegressionClassifier, 'Random Forest', 'Elastic Net']}
+    return graph
+
+
+def test_init(example_graph):
+    comp_graph = ComponentGraph()
+    assert len(comp_graph.component_dict) == 0
+
+    graph = example_graph
+    comp_graph = ComponentGraph(graph)
+    assert len(comp_graph.component_dict) == 6
+
+    expected_order = ['Imputer', 'OneHot_ElasticNet', 'Elastic Net', 'OneHot_RandomForest', 'Random Forest', 'Logistic Regression']
+    assert comp_graph.compute_order == expected_order
+
+
+def test_init_str_components():
+    graph = {'Imputer': ['Imputer'],
+             'OneHot_RandomForest': ['One Hot Encoder', 'Imputer.x'],
+             'OneHot_ElasticNet': ['One Hot Encoder', 'Imputer.x'],
+             'Random Forest': ['Random Forest Classifier', 'OneHot_RandomForest.x'],
+             'Elastic Net': ['Elastic Net Classifier', 'OneHot_ElasticNet.x'],
+             'Logistic Regression': ['Logistic Regression Classifier', 'Random Forest', 'Elastic Net']}
+    comp_graph = ComponentGraph(graph)
+    assert len(comp_graph.component_dict) == 6
+
+    expected_order = ['Imputer', 'OneHot_ElasticNet', 'Elastic Net', 'OneHot_RandomForest', 'Random Forest', 'Logistic Regression']
+    assert comp_graph.compute_order == expected_order
+
+
+def test_invalid_init():
+    invalid_graph = {'Imputer': [Imputer], 'OHE': OneHotEncoder}
+    with pytest.raises(ValueError, match='All component information should be passed in as a list'):
+        ComponentGraph(invalid_graph)
+
+    with pytest.raises(ValueError, match='may only contain str or ComponentBase subclasses'):
+        ComponentGraph({'Imputer': [Imputer(numeric_impute_strategy="most_frequent")], 'OneHot': [OneHotEncoder]})
+
+    graph = {'Imputer': [Imputer(numeric_impute_strategy='constant', numeric_fill_value=0)]}
+    with pytest.raises(ValueError, match='may only contain str or ComponentBase subclasses'):
+        ComponentGraph(graph)
+
+    graph = {'Imputer': ['Imputer', 'Fake'],
+             'Fake': ['Fake Component', 'Estimator'],
+             'Estimator': [ElasticNetClassifier]}
+    with pytest.raises(MissingComponentError):
+        ComponentGraph(graph)
+
+
+def test_init_bad_graphs():
```

Comment:
What about graphs that are partially connected? ie 

```python
graph = {'Imputer': ['Imputer'],
             'OneHot_RandomForest': ['One Hot Encoder', 'Imputer.x'],
             'OneHot_ElasticNet': ['One Hot Encoder', 'Imputer.x'],
             'Random Forest': ['Random Forest Classifier', 'OneHot_RandomForest.x'],
             'Elastic Net': ['Elastic Net Classifier'],
             'Logistic Regression': ['Logistic Regression Classifier', 'Random Forest', 'Elastic Net']}
```
where the 'Elastic Net Classifier' isn't connected to the imputer/OHE? Curious if we should do validation for this, or allow users to make this mistake and have them run into errors later on. This case might be too complicated, but wanted to see what you thought.

Reply:
This is a great point! I have a check to make sure the graph is fully connected, but the example you gave does actually count as being fully connected. I may be able to add a check to ensure there's only one component with no children (in this case catching that the `OneHot_ElasticNet` doesn't connect to anything), since I can't think of any use cases where we would want that behavior!